### PR TITLE
Add optional timeout for wait_until action

### PR DESCRIPTION
**Pull request by jesserockz**
# What does this implement/fix? 

Add optional timeout for wait_until action that will run the next steps in the automation.

Feedback please, perhaps another value `cancel: true` could be added to make it not run the rest of the automations?

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1470

## Test Environment

- [ ] ESP32
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
on_...:
  then:
    - switch.turn_on: ...
    - wait_until:
        condition:
          lambda: "return false;"
        timeout: 2s
    - switch.turn_of: ...
```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by OttoWinter — APPROVED** [review]

**Comment by oxan** [comment]
> Feedback please, perhaps another value cancel: true could be added to make it not run the rest of the automations?

I think it makes sense to be able to distinguish between the condition being true and the timeout being reached, but that can be done in multiple ways. I don't see an issue linked, what's the usecase this aims to solve?

File: esphome/core/base_automation.h [hunk 1]
```diff
@@ -249,6 +257,8 @@ template<typename... Ts> class WaitUntilAction : public Action<Ts...>, public Co
       return;
     }
 
+    this->cancel_timeout("");
```
Comment (OttoWinter) [inline esphome/core/base_automation.h]: Maybe with the new scheduler things have changes, but a timeout with no/empty name at least previously could not be cancelled.

https://github.com/esphome/esphome/blob/262d69308dc45f9ec133ad9aae532dad81efaa8f/esphome/core/component.h#L187

**Comment by jesserockz** [comment]
> I don't see an issue linked, what's the usecase this aims to solve?

I am too lazy to create issues for myself sometimes when I just decide to do something =)

**Comment by jesserockz** [comment]
> I think it makes sense to be able to distinguish between the condition being true and the timeout being reached, but that can be done in multiple ways.


After thinking about it, for now I will leave without being able to check the condition as you could do that with a new `if:` action straight after the `wait_until` to check the same condition, and adding that exact same `then`/`else` logic to the `wait_until` just makes it more complex.

**Pull request merged**